cloud/amazon/s3_lifecycle.py

```diff
@@ -0,0 +1,421 @@
+      required: false
+      default: null
+  state:
+    description:
+      - "Create or remove the lifecycle rule"
+    required: false
+    default: present
+    choices: [ 'present', 'absent' ]
+  status:
+    description:
+      - "If 'enabled', the rule is currently being applied. If 'disabled', the rule is not currently being applied."
+    required: false
+    default: enabled
+    choices: [ 'enabled', 'disabled' ]
+  storage_class:
+    description:
+      - "The storage class to transition to. Currently there is only one valid value - 'glacier'."
+    required: false
+    default: glacier
+    choices: [ 'glacier' ]
+  transition_date:
+    description:
+      - "Indicates the lifetime of the objects that are subject to the rule by the date they will transition to a different storage class. The value must be ISO-8601 format, the time must be midnight and a GMT timezone must be specified. If transition_days is not specified, this parameter is required."
+    required: false
+    default: null
+  transition_days:
+    description:
+      - "Indicates when, in days, an object transitions to a different storage class. If transition_date is not specified, this parameter is required."
+    required: false
+    default: null
+
+extends_documentation_fragment: aws
+'''
+
+EXAMPLES = '''
+# Note: These examples do not set authentication details, see the AWS Guide for details.
+
+# Configure a lifecycle rule on a bucket to expire (delete) items with a prefix of /logs/ after 30 days
+- s3_lifecycle:
+    name: mybucket
+    expiration_days: 30
+    prefix: /logs/
+    status: enabled
+    state: present
+    
+# Configure a lifecycle rule to transition all items with a prefix of /logs/ to glacier after 7 days and then delete after 90 days
+- s3_lifecycle:
+    name: mybucket
+    transition_days: 7
+    expiration_days: 90
+    prefix: /logs/
+    status: enabled
+    state: present
+    
+# Configure a lifecycle rule to transition all items with a prefix of /logs/ to glacier on 31 Dec 2020 and then delete on 31 Dec 2030. Note that midnight GMT must be specified.
+# Be sure to quote your date strings
+- s3_lifecycle:
+    name: mybucket
+    transition_date: "2020-12-30T00:00:00.000Z"
+    expiration_date: "2030-12-30T00:00:00.000Z"
+    prefix: /logs/
+    status: enabled
+    state: present
+    
+# Disable the rule created above
+- s3_lifecycle:
+    name: mybucket
+    prefix: /logs/
+    status: disabled
+    state: present
+    
+# Delete the lifecycle rule created above
+- s3_lifecycle:
+    name: mybucket
+    prefix: /logs/
+    state: absent
+    
+'''
+
+import xml.etree.ElementTree as ET
+import copy
+import datetime
+
+try:
+    import dateutil.parser
+    HAS_DATEUTIL = True
+except ImportError:
+    HAS_DATEUTIL = False
+
+try:
+    import boto.ec2
+    from boto.s3.connection import OrdinaryCallingFormat
```

Comment:
Testing with v1, I had to change this import to:

```python
from boto.s3.connection import OrdinaryCallingFormat, Location
```
I haven't been able to test with v2 because the latest v2 hacking stuff is broken for me.